SI-7886 unsoundness in pattern matcher.

paulp opened the pull request:
I tracked down what was behind the issue described here:

```
  // TODO: fix the illegal type bound in pos/t602 -- type inference
  // messes up before we get here:
  /*override def equals(x$1: Any): Boolean = ...
         // Span[Any] --> Any is not a legal type argument for Span!
         val o5: Option[com.mosol.sl.Span[Any]] =
  */
```

...which led straight to the unsoundness seen in neg/t7886.
It is dangerous to have an expected type of "Any" because
the type system will blithely ignore kind errors, since "Any"
can absorb anything. The consequence in this instance was
that inferring the case constructor for a type like

  `Foo[T <: Bound]`

if done with expected type Any, this would come up with Foo[Any].
I altered it to use expected type Foo[T], which lets the dummy
type parameter survive to carry the bound forward and restores
sense to the inference. The before/after output for -Xprint:patmat
on pos/t602.scala is:

``` diff
15c15
<         if (x1.isInstanceOf[com.mosol.sl.Span[Any]])

---
>         if (x1.isInstanceOf[com.mosol.sl.Span[K]])
17c17
<             <synthetic> val x2: com.mosol.sl.Span[Any] = \
  (x1.asInstanceOf[com.mosol.sl.Span[Any]]: com.mosol.sl.Span[Any]);

---
>             <synthetic> val x2: com.mosol.sl.Span[K] = \
  (x1.asInstanceOf[com.mosol.sl.Span[K]]: com.mosol.sl.Span[K]);
19,20c19,20
<               val low$0: Option[Any] = x2.low;
<               val high$0: Option[Any] = x2.high;

---
>               val low$0: Option[K] = x2.low;
>               val high$0: Option[K] = x2.high;
```

A file in the library depended (needlessly) on the unsoundness.
It was easy to fix but reminds us this may affect existing code.

paulp committed:
SI-7886 unsoundness in pattern matcher.

I tracked down what was behind the issue described here:

  // TODO: fix the illegal type bound in pos/t602 -- type inference
  // messes up before we get here:
  /*override def equals(x$1: Any): Boolean = ...
         // Span[Any] --> Any is not a legal type argument for Span!
         val o5: Option[com.mosol.sl.Span[Any]] =
  */

...which led straight to the unsoundness seen in neg/t7886.
It is dangerous to have an expected type of "Any" because
the type system will blithely ignore kind errors, since "Any"
can absorb anything. The consequence in this instance was
that inferring the case constructor for a type like

  Foo[T <: Bound]

if done with expected type Any, this would come up with Foo[Any].
I altered it to use expected type Foo[T], which lets the dummy
type parameter survive to carry the bound forward and restores
sense to the inference. The before/after output for -Xprint:patmat
on pos/t602.scala is:

15c15
<         if (x1.isInstanceOf[com.mosol.sl.Span[Any]])
---
>         if (x1.isInstanceOf[com.mosol.sl.Span[K]])
17c17
<             <synthetic> val x2: com.mosol.sl.Span[Any] = \
  (x1.asInstanceOf[com.mosol.sl.Span[Any]]: com.mosol.sl.Span[Any]);
---
>             <synthetic> val x2: com.mosol.sl.Span[K] = \
  (x1.asInstanceOf[com.mosol.sl.Span[K]]: com.mosol.sl.Span[K]);
19,20c19,20
<               val low$0: Option[Any] = x2.low;
<               val high$0: Option[Any] = x2.high;
---
>               val low$0: Option[K] = x2.low;
>               val high$0: Option[K] = x2.high;

A file in the library depended (needlessly) on the unsoundness.
It was easy to fix but reminds us this may affect existing code.

paulp commented:
Review @adriaanm

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:
```diff
@@ -322,7 +322,13 @@ trait PatternTypers {
+    private def convertToCaseConstructor(tree: Tree, caseClass: Symbol, ptIn: Type): Tree = {
+      // Unsoundness looms for those who infer type parameters with pt=Any. See SI-7886.
+      val pt = (
+        if (ptIn =:= AnyTpe && caseClass.typeParams.nonEmpty)
```
is there a similar path for Nothing and variance flipped? can pt be wildcard or some such? should we test monomorphicity first assuming it's the common case and thus avoids a =:=?

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:
```diff
@@ -322,7 +322,13 @@ trait PatternTypers {
+      // Unsoundness looms for those who infer type parameters with pt=Any. See SI-7886.
+      val pt = (
+        if (ptIn =:= AnyTpe && caseClass.typeParams.nonEmpty)
+          devWarningResult(s"Evading kind-polymorphic expected type for case constructor of $caseClass")(caseClass.tpe_*)
```
a lot of stuff to read through before getting to the result -- maybe a simple \n can bring solace?

paulp replied:
I have a new fix which fixes all the cases in SI-6680 as well.

paulp replied:
And it avoids testing against Any or anything of the kind, it is somewhere in the realm of "obviously correct."

adriaanm replied:
Great, so shall we close this PR?

paulp replied:
It becomes a challenge. With my better fix, all test pass except two - pos/gadts and run/gadts. Because they're, you know, unsound. I can play with it a little longer. But I think some variation on this PR might be the best I can do, because there are really two different issues at play and this one can (and should) be addressed even if the deeper one isn't.

adriaanm replied:
Ok, I saw your message to scala-internals and agree this is the right PR for the fix -- another overloaded use of Any/Nothing where we should've been using a dedicated type. My bad. Would still be interesting to see the other fix that "breaks" the gadt stuff.

paulp commented:
Wonderful. I'd forgotten all about this one: https://groups.google.com/d/msg/scala-internals/dGQCGQ8dYNs/Yb7WuLQyaMsJ

They rise again.

paulp committed:
SI-6680 unsoundness in gadt typing.

Introduces -Xstrict-inference to deal with the significant
gap between soundness and what presently compiles. I'm hopeful
that it's TOO strict, because it finds e.g. 75 errors compiling
immutable/IntMap.scala, but it might be that bad.

paulp commented:
Re "too strict" in latest commit, I think what has to be done is: correlate the type parameters for the type arguments being inferred with their appearance in any base types upon which inference is being based (n.b. propagateKnownTypes) and only use such type information when the variances match. "When the variances match" is too complicated for me to even describe here, suffice to say I wouldn't take it lightly.

adriaanm commented:
LGTM

#### brainstorming:

I was thinking we may need to consider two directions of the implication arrow in the following description of type inference: assignment of types to type variables => actual type matches expected type.

For a method invocation, we're looking for type arguments that imply the application is well-typed (if you hadn't introduced type parameters, but instead picked some concrete types, the application would've type checked without requiring polymorphism). Versus: you created a value of some polymorphic type, picking actual types for the type parameters involved, we know the value dynamically has a certain type. What does this tell us  about your original choice (what are the implications)?

paulp commented:
I sent an email just now to scala-debate which you probably won't see, so:

```
http://arxiv.org/pdf/1210.5935v1.pdf
"GADT meet subtyping"

While generalized abstract datatypes are now considered well-understood,
adding them to a language with a notion of subtyping comes with a few
surprises. What does it mean for a GADT parameter to be covariant?
The an- swer turns out to be quite subtle and involves new seman-
tic properties of types that raise interesting design ques- tions.
We allow variance annotations in GADT defini- tions, present a sound
and effective algorithm to check such declarations, and describe its
application in a real- world language.

...

What we use in both cases is reasoning of the form1: “if T[β] ≤ α′,
then α′ is of the form T[β′] for some types β′”. We call this an upward
closure property: when we “go up” from a T [β], we only find types that
also have the structure of T. Similarly, for contravariant parameters,
we would need a downward closure property: T is downward-closed if T [β]
≥ α′ entails that α′ is of the form T [β′].

...

In this work, we prove that the notions of upward and downward closures are
the key to a sound variance check for GADT. We start from the formal
development of Si- monet and Pottier [SP07], which provides a general
sound- ness proof for a language with subtyping and a very gen- eral
notion of GADT expressing arbitrary constraints – rather than only type
equalities. By specializing their cor- rectness criterion, we are able
to split it into three smaller criteria, which are simple to implement
in a type-checker. One of them, the most delicate and important one, is
that instances of covariant (respectively contravariant) param- eters
should be upward-closed (resp. downward-closed).
```

paulp commented:
I will continue working on this (sigh) but I will take the lgtm.
